### PR TITLE
Create KG gRPC server

### LINKED ISSUE #15: Create gRPC Server for KG
- [x] Move the proto schema to the kdl-server repo.
- [x] Generate proto code from proto schema to python.
- [x] Using testing assets create a first version of the data pipeline.

### LINKED ISSUE #46: Covert distance to score
- Define a new metric from distance that is a percentage and goes from 0 to 100.

### PR DESCRIPTION
**Pull request by marcvivancos**
Adds an initial version of the gRPC server for the KG.
Closes #15 
Closes #46

### CODE REVIEW

File: kg/src/recommender.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        Gets top paper/repo matches for a given query text.
+        """
+        query_vec = self._compute_query_vector(raw_query_text)
+        query_vec = query_vec.reshape(1, len(query_vec))
+
+        distances = self._compute_cosine_distances(query_vec, self.dataset_vecs)
+
+        df_subset = copy.copy(self.dataset)
+        df_subset['score'] = distances[0]
+        df_subset = df_subset.sort_values(by=['score'], ascending=True).head(n_hits)
```
Comment (IGZVasjaUrbancic) [inline kg/src/recommender.py:93]: This implementation assumes that lowest score is better, whereas I believe the front designs assume the opposite. Equating score with distance here has the potential to quietly break the recommender to return highest-distance hits instead of lowest-distance hits. This code needs revising upon the completion of issue #46. In the meantime, I suggest implementing a placeholder `convert_distance_to_score = lambda x: 1-x` to avoid the risk outlined above.

Comment (IGZVasjaUrbancic) [reply kg/src/recommender.py:93]: Accordingly, to avoid assigning distances to scores, we can use something similar to: 

```
(line 90) scores = [convert_distance_to_score(d) for d in distances]
...
(line 92) df_subset['score'] = scores
```

File: kg/tests/unit/test_recommender.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        data = [["1", "paper", "test title", "test abstract", "test author", 0, "2020-01-05", "http://test", "", ""]]
+        test_dataset = pd.DataFrame(columns=columns, data=data)
+        mock_recommender = Mock()
+        mock_recommender._compute_query_vector.return_value = torch.tensor(np.random.rand(400, 1))
+        mock_recommender._compute_cosine_distances.return_value = torch.tensor(np.random.rand(400, 1))
+        mock_recommender.dataset = test_dataset
+        papers = Recommender.get_top_items(mock_recommender, "test")
+        assert isinstance(papers, outputs.RecommendedList)
+        assert isinstance(papers.items[0], outputs.RecommendedItem)
+        assert papers.items[0].title == "test title"
```
Comment (IGZVasjaUrbancic) [inline kg/tests/unit/test_recommender.py:50]: These assertions all test for useful things, but they test fairly diverse aspects, including the structure of the outputs. In the name of making unit tests more atomic, would it make sense to break this test down into three tests that share a common fixture (lines 40-46) but explicitly test a different aspect each? 

Specifically, some assertions here are testing code that is outside the responsibility of the `Recommender`: line 49, `assert isinstance(papers.items[0], outputs.RecommendedItem)` is in fact testing the `RecommendedList` data class rather than Recommender, so I wonder if they may be better placed elsewhere (`test_outputs.py`). Alternatively, simply separating this out into three tests and giving each of them an explicit name for what exactly it is testing may be enough. 

This is not essential, it's just a suggestion for improvement for long-term maintainability if you agree that it would help and if time permits.

File: kg/tests/unit/test_tools_assets.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+class TestAssetLoader:
+
+    def test_asset_loader_mismatch_lengths(self):
+        with pytest.raises(AssertionError):
+            assets_mock = Mock()
+            assets_mock.dataset = [1, 2, 3, 4]
+            assets_mock.vectors = [1, 2, 3]
+
+            AssetLoader._asset_checks(assets_mock)
```
Comment (IGZVasjaUrbancic) [inline kg/tests/unit/test_tools_assets.py:15]: I suggest adding a happy-path test here, i.e. a case where AssetLoader is able to successfully load the provided dataset and vectors.

In addition, that's a minute detail, but would it make sense to match the actual expected input types (i.e. np array of floats for vectors, data frame for dataset) in the test case? They all implement __len__, so at the moment it won't make a practical difference, but if in the future we decide for example to create our own data class instead of a data frame for `dataset`, having the actual class of the asset inside this test would be useful to make sure it behaves correctly.

File: kg/tests/unit/test_recommender.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import outputs
+import pandas as pd
+import pytest
+import torch
+from recommender import Recommender
+
+
+class TestRecommender:
+
+    def test_preprocess(self):
```
Comment (IGZVasjaUrbancic) [inline kg/tests/unit/test_recommender.py:13]: I suggest a test name that explicitly states what exactly it tests for, e.g. `test_that_preprocess_removes_newline_characters`

File: kg/src/recommender.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        Tokenizes input text
+        """
+        tokens = self.tokenizer(text, truncation=True, max_length=512, return_tensors='pt')
+        return tokens
+
+    @staticmethod
+    def _compute_cosine_distances(vectors_1: np.ndarray, vectors_2: np.ndarray) -> np.ndarray:
+        """
+        Computes pairwise cosine distances between vectors_1 (array) and vectors_2 (array).
+        """
```
Comment (IGZVasjaUrbancic) [inline kg/src/recommender.py:62]: It may be useful to specify the expected shapes of input and output vectors in the docstring, as it is intended to be used within the recommender. `(1, vector_dimensions)` representing the query vector in `vectors_1` ,  `(N_HITS, vector_dimensions)` representing the dataset vectors for `vectors_2`, and `(N_HITS, 1)` for output (or something similar, not sure exactly).

File: kg/src/outputs.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    def __ge__(self, other):
+        return self.score >= other.score
+
+    def __le__(self, other):
+        return self.score <= other.score
+
+
+@dataclass(init=False)
+class RecommendedList:
```
Comment (IGZVasjaUrbancic) [inline kg/src/outputs.py:64]: Maybe a brief docstring here as well?

File: kg/.flake8 [hunk 7]
```diff
@@ -0,0 +1,5 @@
+[flake8]
+doctest = True
+max-line-length = 115
+exclude = tests/*,.git,__pycache__,proto,.github/
```
Comment (IGZVasjaUrbancic) [inline kg/.flake8:4]: Are tests excluded on purpose?

**Review by IGZVasjaUrbancic — CHANGES_REQUESTED** [review]
Very nice work! I think one item really needs changing before merging, which is relating to the comment on scores and distances in `Recommender.get_top_items()`. Other comments I made are suggestions for improvement but are not essential to address before merging. Thank you Marc, great work!

File: kg/.editorconfig [hunk 8]
```diff
@@ -0,0 +1,10 @@
+[*]
+indent_style = space
+indent_size = 4
+insert_final_newline = true
+trim_trailing_whitespace = true
+end_of_line = lf
+charset = utf-8
+
+[*.py]
+max_line_length = 119
```
Comment (angelsanchez) [inline kg/.editorconfig:10]: Why the max_line_length (119) is different than .flake8 (115) configuration?

File: kg/README.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+## Run server
+
+Install the dependencies using `pipenv`:
+```
+pipenv install
+```
+
+and then run:
+```
+python src/app.py
```
Comment (angelsanchez) [inline kg/README.md:32]:
```suggestion
pipenv run python src/app.py
```
the python interpreter should be the located inside the virtual env instead of the installed in the path.

File: kg/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+### Unit tests
+To run the unit tests you have to install the development dependencies:
+```
+pipenv install --dev
+```
+
+and then run:
+```
+python -m pytest -m "not int"
```
Comment (angelsanchez) [inline kg/README.md:48]:
```suggestion
pipenv run pytest -m "not int"
```

File: kg/README.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+# flake8 checks code conventions (pep8...) and code style, docs...
+flake8
+# mypy is not fully fuctional but it's useful, it simulates static typing
+mypy [file-to-analyze]
+```
+
+### Integration tests
+To run the integration tests you have to set up a server:
+```
+python src/app.py
```
Comment (angelsanchez) [inline kg/README.md:63]:
```suggestion
pipenv run python src/app.py
```

File: kg/README.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+```
+and then run:
+```
+python -m pytest int
```
Comment (angelsanchez) [inline kg/README.md:67]:
```suggestion
pipenv run pytest int
```

File: kg/src/recommender.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+import numpy as np
+import pandas as pd
+import torch
+import transformers
+from scipy import spatial
+
+import outputs
+
+log = logging.getLogger("Recommender")
```
Comment (angelsanchez) [inline kg/src/recommender.py:12]: Consider to use a class member like other classes, e.g.
```
self.log = logging.getLogger("KGApp")
```

**Review by angelsanchez — CHANGES_REQUESTED** [review]

**Review by angelsanchez — APPROVED** [review]

**Review by IGZVasjaUrbancic — APPROVED** [review]
All good :+1:  Thank you Marc!
